Core/Core/Utils/ServiceRegistry.cs

```diff
@@ -0,0 +1,93 @@
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace AeroGear.Mobile.Core.Utils
+{
+    ///<summary>
+    ///This class maintains mappings for <see cref="MobileCore"/> of platform specific implementaityons of <see cref="IServiceModule"/>
+    ///</summary>
+    sealed class ServiceRegistry
+    {
+        private static Object registryLock = new Object();
+
+        private readonly ILogger logger;
+        private readonly IPlatformInjector injector;
+        private List<Type> dependencyTypes;
+        
+        internal ServiceRegistry(ILogger logger, IPlatformInjector injector) {
+            this.logger = logger;
+            this.injector = injector;
+        }
+
+        /// <summary>
+        /// This method scans all assemblies for classes annotated with <see cref="ServiceAttribute"/>.
+        /// </summary>
+        /// <param name="injector">The <see cref="IPlatformInjector"/> to use to get access to platform resources.</param>
+        internal async Task RegisterAssemblies()
+        {
+            await Task.Run(() => {
+                lock (registryLock)
+                {
+                    if (dependencyTypes != null) {
+                        return;//Short circuit as we've already calculated the dependency tree;
+                    }
+                    dependencyTypes = new List<Type>();
+                    Assembly[] assemblies = injector.GetAssemblies();
+
+                    Type targetAttrType = typeof(ServiceAttribute);
+
+                    // Don't use LINQ for performance reasons
+                    // Naive implementation can easily take over a second to run
+                    foreach (Assembly assembly in assemblies)
+                    {
+                        Attribute[] attributes;
+                        try
+                        {
+                            attributes = assembly.GetCustomAttributes(targetAttrType).ToArray();
+                        }
+                        catch (System.IO.FileNotFoundException exception)
+                        {
+                            // Sometimes the previewer doesn't actually have everything required for these loads to work
+                            logger.Error($"Could not load assembly: {assembly.FullName} for Attibute {targetAttrType.FullName} | Some renderers may not be loaded", exception);
+                            continue;
+                        }
+
+                        if (attributes.Length == 0)
+                            continue;
+
+                        foreach (ServiceAttribute attribute in attributes)
+                        {
+                            if (!dependencyTypes.Contains(attribute.Implementor))
+                            {
+                                dependencyTypes.Add(attribute.Implementor);
+                            }
+                        }
+                    }
+                }
```

Comment:
do we know roughly how long the look up will take?

Reply:
It takes about a second.

Reply:
If we make this working, I think we could even make it faster by filtering on the assembly name. For example, we could call all our assembly something like "aerogear.xxx" and lookup only into assembly named that way.